### ISSUE TITLE
grad(softplus) has wrong value at zero

### ISSUE THREAD

**Issue by shoyer** [google/jax#2107]
`jax.grad(jax.nn.softplus)(0.0)` evaluates to 0.0, which is definitely wrong -- the right answer is 0.5.

This is easy to visualize:
```python
import matplotlib.pyplot as plt
x = jax.numpy.linspace(-2, 2, num=101)
plt.plot(x, jax.vmap(jax.grad(jax.nn.softplus))(x))
```
![image](https://user-images.githubusercontent.com/1217238/73335588-b4ee6680-4224-11ea-8061-3982ea777051.png)

**Comment by hawkinsp** [comment]
`logaddexp` is implemented as:
```
  amax = lax.max(x1, x2)
  delta = lax.sub(x1, x2)
  return lax.select(isnan(delta),
                    lax.add(x1, x2),  # NaNs or infinities of the same sign.
                    lax.add(amax, lax.log1p(lax.exp(-lax.abs(delta)))))
```
The derivative of the implementation, especially the `abs` and `max` seem problematic.

We'd probably do best to either make `logaddexp` primitive or to add a custom JVP rule for it.

**Comment by ibab** [comment]
We have run into this on a real project so I think it would be good if we could fix this as soon as possible.

How about this version?
```python
def logaddexp2(x1, x2):
  amax = lax.max(x1, x2)
  amin = lax.min(x1, x2)
  delta = amin - amax
  return lax.select(jnp.isnan(delta),
                    lax.add(x1, x2),
                    lax.add(amax, lax.log1p(lax.exp(delta)))
```
It seems to be slightly slower than the original:
``` python
xs = np.linspace(-5, 5, 1000001)
f1 = jax.jit(jax.vmap(jax.grad(logaddexp)))
f2 = jax.jit(jax.vmap(jax.grad(logaddexp2)))
ys = np.zeros_like(xs)
f1(xs, ys)
f2(xs, ys)
%timeit f1(xs, ys)
%timeit f2(xs, ys)
```
```
100 loops, best of 3: 5.79 ms per loop
100 loops, best of 3: 6.35 ms per loop
```
But the slowdown is not too bad, should I make a PR with this version as an interim solution?

**Comment by joaogui1** [comment]
I'm attempting a solution that changes lax.max grad, #2195

**Comment by mattjj** [comment]
@ibab That seems reasonable, though I'm not sure we can rely on XLA preserving NaN semantics under a `jit` on all backends.

@joaogui1 Thanks for digging in! I'm not sure we should couple fixing this issue to changing our derivative for `lax.max`, though. We can discuss what the best policy for `lax.max` is, but this issue seems to be crying out for more direct solution, like a custom JVP rule.

**Comment by mattjj** [comment]
@ibab re: NaN semantics, I have learned that XLA _will_ respect NaN semantics (and while LLVM on CPU might not, JAX currently sets flags so that it does). So your solution seems viable.

We're looking into this more...

**Comment by joaogui1** [comment]
@mattjj okss, but I think we should take a look at `lax.max` later, as right now it's gradient is not compatible with `lax.abs` gradient
